### PR TITLE
Avoid loading style source if its layer(s) are not shown

### PR DESCRIPTION
**Pull request by brunoabinader**
Given https://github.com/mapbox/mapbox-gl-styles/blob/master/styles/satellite-hybrid-v8.json as example of a style containing 2+ sources, if all layers related to a specific source are either disabled (e.g. due to class selections) or invisible, we should avoid _loading_ the style source entirely. This saves processing time and network bandwidth.

When the layers related to a specific style source are re-enabled, the burden of loading the style source could be reduced by using a cache file source.

/cc @mapbox/gl @mapbox/mobile

### CODE REVIEW

File: src/mbgl/style/style_layer.hpp [hunk 1]
```diff
@@ -45,6 +45,8 @@ class StyleLayer {
     // Returns true if any paint properties have active transitions.
     virtual bool recalculate(const StyleCalculationParameters&) = 0;
 
+    virtual bool isHidden() const = 0;
```
Comment (jfirebaugh) [inline src/mbgl/style/style_layer.hpp]: Rename this to `isVisible` and reverse the boolean, for consistency with "visibility" as from the style spec.

**Comment by jfirebaugh** [comment]
Instead of introducing more state, can we use `Source::enabled` for this?

**Comment by brunoabinader** [comment]
> Instead of introducing more state, can we use Source::enabled for this?

Indeed, _enabling_ the source only if the layer is visible in `Style::recalculate()` is nicer than comparing strings in `Style::loadSources()`. However, we still require `Source::isLoading()` to avoid double requests and `StyleLayer::isVisible()` because each StyleLayer-derived class calculates its visibility in a different way. I'll update the code accordingly.

**Comment by jfirebaugh** [comment]
There are some corner cases where a layer with a zero-opacity color still renders something, e.g. symbols with halo but transparent fill, background-pattern, outline-only fill layers. Can we limit the check to `opacity > 0`? Also we should add checks of the `*-visibility` layout property, which is the most common way to hide a layer completely.

**Comment by jfirebaugh** [comment]
Can we make `layer.isVisible()` equivalent to `layer.passes != RenderPass::None`?

File: src/mbgl/map/source.cpp [hunk 2]
```diff
@@ -52,6 +52,14 @@ bool Source::isLoaded() const {
     return true;
 }
 
+bool Source::isLoading() const {
+    if (loaded) {
+        return false;
+    }
+
+    return req.operator bool();
```
Comment (jfirebaugh) [inline src/mbgl/map/source.cpp]: Should be a one liner: `return !loaded && req.operator bool();`

**Comment by brunoabinader** [comment]
> Can we limit the check to opacity > 0?
> Also we should add checks of the *-visibility layout property, which is the most common way to hide a layer completely.

This is used in `Style::recalculate()` now:
```
bool StyleLayer::needsRendering() const {
    return passes != RenderPass::None && visibility != VisibilityType::None;
}
```

I had to fix/modify some style unit tests to adapt to the new scheme for source loading - instead of immediately load sources upon style JSON parsing, we wait to check whether a layer actually needs it.

File: src/mbgl/map/source.cpp [hunk 3]
```diff
@@ -39,9 +39,7 @@ Source::Source() {}
 Source::~Source() = default;
 
 bool Source::isLoaded() const {
-    if (!loaded) {
```
Comment (jfirebaugh) [inline src/mbgl/map/source.cpp]: Is this change necessary? I would expect `Source:isLoaded()` to return true in the case that at some zN, it was enabled and loaded, and then at zM the only layer that used it becomes hidden and the source was therefore disabled.

File: src/mbgl/style/style.cpp [hunk 4]
```diff
@@ -66,6 +66,14 @@ void Style::setJSON(const std::string& json, const std::string&) {
     loaded = true;
 }
 
+void Style::loadSources() {
+    for (auto& source : sources) {
+        if (source->loaded) continue;
```
Comment (jfirebaugh) [inline src/mbgl/style/style.cpp]: I think this is clearer if the conditions are all in the same direction:

```
if (source->enabled && !source->loaded && !source->isLoading()) {
    source->load();
}
```

File: src/mbgl/style/style.cpp [hunk 5]
```diff
@@ -166,6 +173,8 @@ void Style::cascade() {
     for (const auto& layer : layers) {
         layer->cascade(parameters);
     }
+
+    loadSources();
```
Comment (jfirebaugh) [inline src/mbgl/style/style.cpp]: I think `loadSources()` is not necessary here -- `cascade` doesn't change `source->enabled` so it can never trigger a source being needed to load.

File: src/mbgl/map/source.cpp (hunk 3 above)
Comment (brunoabinader) [inline src/mbgl/map/source.cpp]: Agreed about `enabled`, but I believe we also should check if there is a pending request, which indicates the source is in loading state.

Comment (jfirebaugh) [inline src/mbgl/map/source.cpp]: Doesn't `req` imply `!loaded` though? `loaded` only gets set to true after all the requests are successful.

Comment (brunoabinader) [inline src/mbgl/map/source.cpp]: Yes :dart: Good catch, indeed!
